### PR TITLE
chore: Pre-build DevWorkspace templates for devfile v2

### PR DESCRIPTION
**Pull request by vinokurig**
Signed-off-by: Igor Vinokur <ivinokur@redhat.com>

### What does this PR do?
Generate DW templates at build time and put them into `/usr/local/apache2/htdocs/devfiles/v2/che-thei/` folder in the che-devfile-registry image.

### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/19718

### How to test this PR?

1. Build and run the `che-devfile-registry` image.
2. Go to `<devfile-registry URL>/devfiles/v2/che-theia/`.
3. See: List of the generated tamplates.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: build/dockerfiles/Dockerfile [hunk 1]
```diff
@@ -19,6 +19,13 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+FROM docker.io/node:12.20.1-alpine3.12 as dwtamplates
```
Comment (benoitf) [inline build/dockerfiles/Dockerfile:22]: if you add something there, is needs to be added into rhel.Dockerfile as well

Comment (benoitf) [inline build/dockerfiles/Dockerfile:22]: `dwtamplates` --> `dwtemplates`

File: build/scripts/generate_devworkspace_templates.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+mkdir /build/out/
+for dir in /build/devfiles/*/
+do
+  devfile=$(cat ${dir}meta.yaml | yq -r '."links"."v2"')
+  if [ $devfile != null ]; then
+    dir=${dir%/}
+    dir=/build/out/${dir##*/}
+    mkdir ${dir}
+    npx @eclipse-che/che-theia-devworkspace-handler --devfile-url:$devfile --output-file:${dir}/meta.yaml
```
Comment (benoitf) [inline build/scripts/generate_devworkspace_templates.sh:19]: the output should be with a different name than meta.yaml as we're generating a file that is containing a DevWorkspace + DevWorkspaceTemplate

Comment (benoitf) [reply build/scripts/generate_devworkspace_templates.sh:19]: it looks overkill to install jq and yq just to know if there is v2.links in the meta.yaml while probably only a grep expression is enough

File: build/scripts/generate_devworkspace_templates.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+#
+# SPDX-License-Identifier: EPL-2.0
+#
+
+mkdir /build/out/
+for dir in /build/devfiles/*/
+do
+  devfile=$(cat ${dir}meta.yaml | yq -r '."links"."v2"')
+  if [ $devfile != null ]; then
```
Comment (benoitf) [inline build/scripts/generate_devworkspace_templates.sh:15]: I'm not sure != null is bash

File: build/dockerfiles/Dockerfile (hunk 1 above)
Comment (vinokurig) [reply build/dockerfiles/Dockerfile]: done

File: build/scripts/generate_devworkspace_templates.sh (hunk 2 above)
Comment (vinokurig) [reply build/scripts/generate_devworkspace_templates.sh]: done

Comment (svor) [reply build/scripts/generate_devworkspace_templates.sh]: >the output should be with a different name than meta.yaml as we're generating a file that is containing a DevWorkspace + DevWorkspaceTemplate

@benoitf can we just name it `devWorkspace.yaml`?

File: build/dockerfiles/Dockerfile [hunk 4]
```diff
@@ -19,6 +19,13 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+FROM docker.io/node:12.20.1-alpine3.12 as dwtemplates
+RUN apk add --no-cache bash && npm install -g @eclipse-che/che-theia-devworkspace-handler
+COPY ./build/scripts /build/
+COPY ./devfiles /build/devfiles
+WORKDIR /build/
+RUN ./generate_devworkspace_templates.sh
+
```
Comment (benoitf) [inline build/dockerfiles/Dockerfile:28]: we're still missing instructions in https://github.com/eclipse-che/che-devfile-registry/blob/main/build/dockerfiles/rhel.Dockerfile

File: build/scripts/generate_devworkspace_templates.sh (hunk 2 above)
Comment (benoitf) [reply build/scripts/generate_devworkspace_templates.sh]: yes why not it's better than meta.yaml

File: build/dockerfiles/Dockerfile (hunk 1 above)
Comment (vinokurig) [reply build/dockerfiles/Dockerfile]: done

File: build/scripts/generate_devworkspace_templates.sh (hunk 2 above)
Comment (vinokurig) [reply build/scripts/generate_devworkspace_templates.sh]: done

File: build/dockerfiles/Dockerfile (hunk 4 above)
Comment (vinokurig) [reply build/dockerfiles/Dockerfile:28]: Added in the latest fixup

File: build/dockerfiles/Dockerfile [hunk 5]
```diff
@@ -19,6 +19,13 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+FROM docker.io/node:12.20.1-alpine3.12 as dwtemplates
+RUN apk add --no-cache bash && npm install -g @eclipse-che/che-theia-devworkspace-handler
```
Comment (benoitf) [inline build/dockerfiles/Dockerfile:23]: the version of the library @eclipse-che/che-theia-devworkspace-handler should be added to be sure to have a reproducible build

File: build/dockerfiles/Dockerfile [hunk 6]
```diff
@@ -19,6 +19,13 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+FROM docker.io/node:12.20.1-alpine3.12 as dwtemplates
```
Comment (benoitf) [inline build/dockerfiles/Dockerfile:22]: should we use a recent LTS version there ? like nodejs 16

File: build/dockerfiles/rhel.Dockerfile [hunk 7]
```diff
@@ -52,6 +52,16 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+
+FROM registry.access.redhat.com/ubi8/nodejs-12:1-102 as dwtemplates
```
Comment (benoitf) [inline build/dockerfiles/rhel.Dockerfile:56]: probably should use a more recent version (12 is no longer an active LTS version)

File: build/dockerfiles/Dockerfile (hunk 5 above)
Comment (benoitf) [inline build/dockerfiles/Dockerfile:23]: if we don't need bash, we can have the npm install command in the script

File: build/dockerfiles/rhel.Dockerfile [hunk 8]
```diff
@@ -52,6 +52,16 @@ RUN ./index.sh > /build/devfiles/index.json
...
+USER 0
+
+RUN npm install -g @eclipse-che/che-theia-devworkspace-handler
```
Comment (benoitf) [inline build/dockerfiles/rhel.Dockerfile:59]: I think it won't work in brew environment (cc @nickboldt for downstreaming)

File: build/dockerfiles/Dockerfile (hunk 5 above)
Comment (vinokurig) [reply build/dockerfiles/Dockerfile]: done

File: build/dockerfiles/Dockerfile (hunk 6 above)
Comment (vinokurig) [reply build/dockerfiles/Dockerfile]: Changed to `node:lts-alpine3.14`

File: build/dockerfiles/rhel.Dockerfile (hunk 7 above)
Comment (vinokurig) [reply build/dockerfiles/rhel.Dockerfile]: Changed to `registry.access.redhat.com/ubi8/nodejs-14-minimal:1-31`

File: build/dockerfiles/Dockerfile (hunk 5 above)
Comment (vinokurig) [reply build/dockerfiles/Dockerfile]: done

File: build/dockerfiles/rhel.Dockerfile (hunk 8 above)
Comment (vinokurig) [reply build/dockerfiles/rhel.Dockerfile]: Wouldn't it be overridden here: https://github.com/redhat-developer/codeready-workspaces/blob/crw-2-rhel-8/dependencies/che-devfile-registry/build/dockerfiles/Dockerfile?

Comment (nickboldt) [reply build/dockerfiles/rhel.Dockerfile]: yes, CRW devfile registry is a fork. There's no sync between che devfile reg and the crw one.

Therefore you need to manually apply these changes to downstream. But you cannot just `npm install` in brew unless either:
* you prebuild the bits in jenkins as a tarball then unpack them during the brew build as if doing an offline build, or
* use cachito to fetch npm dependencies https://issues.redhat.com/browse/CRW-2444

File: build/dockerfiles/rhel.Dockerfile (hunk 7 above)
Comment (benoitf) [reply build/dockerfiles/rhel.Dockerfile]: why not nodejs-16, 14 is no longer active as well ?
https://nodejs.org/en/about/releases/#releases

Comment (benoitf) [reply build/dockerfiles/rhel.Dockerfile]: (also you pickup nodejs lts for alpine which is for now v16.13.0.)

File: build/dockerfiles/Dockerfile [hunk 9]
```diff
@@ -19,6 +19,12 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+FROM docker.io/node:lts-alpine3.14 as dwtemplates
```
Comment (benoitf) [inline build/dockerfiles/Dockerfile:22]:
```suggestion
FROM docker.io/node:16.13.0-alpine3.14 as dwtemplates
```

lts is a rolling tag and we should use only 'fixed' tag so in 6months or one year it's still using the same image

File: build/dockerfiles/rhel.Dockerfile [hunk 10]
```diff
@@ -52,6 +52,15 @@ RUN ./index.sh > /build/devfiles/index.json
 RUN ./list_referenced_images.sh devfiles > /build/devfiles/external_images.txt
 RUN chmod -R g+rwX /build/devfiles
 
+
+FROM registry.access.redhat.com/ubi8/nodejs-14-minimal:1-31 as dwtemplates
```
Comment (benoitf) [inline build/dockerfiles/rhel.Dockerfile:56]:
```suggestion
FROM registry.access.redhat.com/ubi8/nodejs-16-minimal:1-6 as dwtemplates
```

File: build/scripts/generate_devworkspace_templates.sh [hunk 11]
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh
+#
+# Copyright (c) 2019-2021 Red Hat, Inc.
```
Comment (benoitf) [inline build/scripts/generate_devworkspace_templates.sh:3]:
```suggestion
# Copyright (c) 2021 Red Hat, Inc.
```

File: build/scripts/generate_devworkspace_templates.sh [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+npm install -g @eclipse-che/che-theia-devworkspace-handler@0.0.1-1637592995
+mkdir /build/out/
+for dir in /build/devfiles/*/
+do
+  devfile=$(cat ${dir}meta.yaml | grep "v2:")
+  if [ -n "$devfile" ]; then
+    dir=${dir%/}
+    dir=/build/out/${dir##*/}
+    mkdir ${dir}
+    npx @eclipse-che/che-theia-devworkspace-handler --devfile-url:${devfile##*v2: } --output-file:${dir}/devWorkspace.yaml
```
Comment (benoitf) [inline build/scripts/generate_devworkspace_templates.sh:20]:
```suggestion
    npx @eclipse-che/che-theia-devworkspace-handler --devfile-url:${devfile##*v2: } --output-file:${dir}/devworkspace.yaml
```
maybe it's better to not have camelCase (we only have lowercase until now) or to use 'dev-workspace.yaml'

**Comment by benoitf** [comment]
Thank you for all the fixes

I still have a question:

- how do we get the link to this dev workspace templates ? I don't see meta.yaml being updated with the link to che-theia devWorkspace templates for example ?
- What are the images of che-theia used ? for example when we cut a release like 7.40.0 what devworkspace.yaml file will contain ? as it should reference che-theia:7.40.0 images in that case ?

**Comment by svor** [comment]
The build of image was interrupted by:
```
Run `npm audit` for details.
npm notice 
npm notice New patch version of npm available! 8.1.0 -> 8.1.4
npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.1.4
npm notice Run npm install -g npm@8.1.4 to update!
npm notice 
Need to install the following packages:
  @eclipse-che/che-theia-devworkspace-handler
Ok to proceed? (y)
```
Would it be possible to install che-theia-devworkspace-handler without interrupting?

**Comment by vinokurig** [comment]
@svor
>The build of image was interrupted by:
>
>
>```
>Run `npm audit` for details.
>npm notice 
>npm notice New patch version of npm available! 8.1.0 -> 8.1.4
>npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.1.4
>npm notice Run npm install -g npm@8.1.4 to update!
>npm notice 
>Need to install the following packages:
>  @eclipse-che/che-theia-devworkspace-handler
>Ok to proceed? (y)
>```
>Would it be possible to install che-theia-devworkspace-handler without interrupting?

What command did you use to build the image?

**Comment by svor** [comment]
After the discussion with @benoitf we decided to add the link to this dev workspace templates into `meta.yaml` at build time   
and put generated files into devfile folder. For example final `meta.yaml` file for Go devfile should be like:

```json
{
  "displayName": "Go",
  "description": "Stack with Go 1.14",
  "tags": [
    "Debian",
    "Go"
  ],
  "icon": "/images/go.svg",
  "links": {
    "v2": "https://github.com/che-samples/golang-echo-example/tree/devfile2",
    "devWorkspaces": {
       "eclipse/che-theia/latest": "/devfiles/go/devworkspace-che-theia-latest.yaml"
       "eclipse/che-theia/next": "/devfiles/go/devworkspace-che-theia-next.yaml"
    },
    "self": "/devfiles/go/devfile.yaml"
  }
}
```
The only difference between `eclipse/che-theia/next` and `eclipse/che-theia/latest` is that for `eclipse/che-theia/latest` we have to use tags of images taken from https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/eclipse/che-theia/latest/

**Comment by benoitf** [comment]
I think we should call the library two times as there is already a --editor flag

1. `... --editor:eclipse/che-theia/next --devfile-url:....`
2. `... --editor:eclipse/che-theia/latest --devfile-url:...`

**Comment by nickboldt** [comment]
seems reasonable. Please remember to submit a PR for downstreaming this into https://github.com/redhat-developer/codeready-workspaces/tree/crw-2-rhel-8/dependencies/che-devfile-registry/devfiles for CRW 2.14

**Comment by vinokurig** [comment]
@svor @benoitf
> After the discussion with @benoitf we decided to add the link to this dev workspace templates into `meta.yaml` at build time   
and put generated files into devfile folder. For example final `meta.yaml` file for Go devfile should be like:
>
>```json
>{
>  "displayName": "Go",
>  "description": "Stack with Go 1.14",
>  "tags": [
>    "Debian",
>    "Go"
>  ],
>  "icon": "/images/go.svg",
>  "links": {
>    "v2": "https://github.com/che-samples/golang-echo-example/tree/devfile2",
>    "devWorkspaces": {
>       "eclipse/che-theia/latest": "/devfiles/go/devworkspace-che-theia-latest.yaml"
>       "eclipse/che-theia/next": "/devfiles/go/devworkspace-che-theia-next.yaml"
>    },
>    "self": "/devfiles/go/devfile.yaml"
>  }
>}
>```
>The only difference between `eclipse/che-theia/next` and `eclipse/che-theia/latest` is that for `eclipse/che-theia/latest` we have to use tags of images from https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/eclipse/che-theia/latest/

Json with slashes in a key name are not valid.
What about such structure:
```json 
{
  "links": {
    "v2": "https://github.com/che-samples/golang-echo-example/tree/devfile2",
    "devWorkspaces": {
       "eclipse": {
         "che-theia": {
           "latest": "/devfiles/go/devworkspace-che-theia-latest.yaml",
           "next": "/devfiles/go/devworkspace-che-theia-next.yaml"
        }
      }
    },
    "self": "/devfiles/go/devfile.yaml"
  }
}
```

**Comment by benoitf** [comment]
@vinokurig could you tell why slash in keys are invalid ?
we've using tons of slash in package.json dependencies

**Comment by vinokurig** [comment]
@benoitf
* From the convention: https://google.github.io/styleguide/jsoncstyleguide.xml?showone=Property_Name_Format#Property_Name_Format:
`Subsequent characters can be a letter, a digit, an underscore, or a dollar sign.`
* `yq` fails when parsing such key
* An online JSON validator fails parsing such json

**Comment by benoitf** [comment]
I'm not sure if you're talking about json or yaml

but either in JSON or Yaml I'm able to work with key/slash properties

I think you mean yaml as we've meta.yaml
```
/ # cat /project/foo.yaml
description: "Stack with Go 1.14"
displayName: Go
icon: /images/go.svg
links:
  devWorkspaces:
    eclipse/che-theia/latest: /devfiles/go/devworkspace-che-theia-latest.yaml
    eclipse/che-theia/next: /devfiles/go/devworkspace-che-theia-next.yaml
  self: /devfiles/go/devfile.yaml
  v2: "https://github.com/che-samples/golang-echo-example/tree/devfile2"
tags:
  - Debian
  - Go


/ # yq '.icon' /project/foo.yaml
"/images/go.svg"
/ # yq '.links.devWorkspaces' /project/foo.yaml
{
  "eclipse/che-theia/latest": "/devfiles/go/devworkspace-che-theia-latest.yaml",
  "eclipse/che-theia/next": "/devfiles/go/devworkspace-che-theia-next.yaml"
}
/ # yq '.links.devWorkspaces."eclipse/che-theia/next"' /project/foo.yaml
"/devfiles/go/devworkspace-che-theia-next.yaml"
```

And it works on online validator as well

image

image

**Comment by vinokurig** [comment]
I am sorry for confusing by mixing JSON and YAML.
`yq` fails when I pass `yq '.links.devWorkspaces |= . + {eclipse/che-theia/latest: "<some value>"}' "${meta}")`
I couldn't find any conventions for YAML but it looks like `yq` uses `jq` under the hood so may be it validates the key with JSON rules.

**Comment by benoitf** [comment]
```
/ # yq '.links.devWorkspaces | . + {"i/can/use/slash/there" : "value"} ' /project/foo.yaml
{
  "eclipse/che-theia/latest": "/devfiles/go/devworkspace-che-theia-latest.yaml",
  "eclipse/che-theia/next": "/devfiles/go/devworkspace-che-theia-next.yaml",
  "i/can/use/slash/there": "value"
}
```

**Comment by vinokurig** [comment]
@benoitf Thank you for the example. I forgot to add ""

**Comment by benoitf** [comment]
FYI shellcheck PR check is failing

File: build/dockerfiles/rhel.Dockerfile [hunk 13]
```diff
@@ -84,6 +93,7 @@ WORKDIR /var/www/html
 RUN mkdir -m 777 /var/www/html/devfiles
 COPY .htaccess README.md /var/www/html/
 COPY --from=builder /build/devfiles /var/www/html/devfiles
+COPY --from=dwtemplates /build/out /var/www/html/devfiles/v2/che-theia/
```
Comment (benoitf) [inline build/dockerfiles/rhel.Dockerfile:96]: my understanding was that the templates should go in the go sample folder if it's for go for example (and not che-theia) ?

File: build/scripts/index.sh [hunk 14]
```diff
@@ -20,5 +20,10 @@ for meta in "${metas[@]}"; do
     # Ignore double quotes warning for yq expression
     # shellcheck disable=SC2016,SC2094
     cat <<< "$(yq --arg metadir "${META_DIR}" '.links |= . + {self: "/\($metadir)/devfile.yaml" }' "${meta}")"  > "${meta}"
+    if [ "$(yq '.links.v2' "${meta}")" != "null" ]; then
+      cat <<< "$(yq '.links.devWorkspaces |= . +
+      {"eclipse/che-theia/latest": "/devfiles/go/devworkspace-che-theia-latest.yaml",
```
Comment (benoitf) [inline build/scripts/index.sh:25]: I don't see how it can be hardcoded to /devfiles`/go/` folder

File: build/scripts/index.sh [hunk 15]
```diff
@@ -20,5 +20,10 @@ for meta in "${metas[@]}"; do
...
+      "eclipse/che-theia/next": "/devfiles/go/devworkspace-che-theia-next.next"}' "${meta}")" > "${meta}"
```
Comment (benoitf) [inline build/scripts/index.sh:26]: extension should be .yaml probably

File: build/dockerfiles/rhel.Dockerfile (hunk 13 above)
Comment (vinokurig) [reply build/dockerfiles/rhel.Dockerfile:96]: it was commited mistakenly, reverted the fixup

File: build/scripts/index.sh (hunk 14 above)
Comment (vinokurig) [reply build/scripts/index.sh]: it was commited mistakenly, reverted the fixup

File: build/scripts/index.sh (hunk 15 above)
Comment (vinokurig) [reply build/scripts/index.sh]: it was commited mistakenly, reverted the fixup

**Comment by vinokurig** [comment]
@svor @benoitf 
* The `meta.yaml` file is updated with `devWorkspaces` links.
* DevWorkspace templates are generated for both `next` and `latest` versions in the extension directory.
* shellcheck successfully passed.

**Comment by benoitf** [comment]
I tested but I have as links

```
links": {
"v2": "https://github.com/che-samples/java-guestbook/tree/devfilev2",
"self": "/devfiles/java-mongo/devfile.yaml",
"devWorkspaces": {
"eclipse/che-theia/latest": "devfiles/java-mongo/devworkspace-che-theia-latest.yaml",
"eclipse/che-theia/next": "devfiles/java-mongo/devworkspace-che-theia-next.yaml"

```

so I suppose there is a missing / before devfiles ?

File: build/scripts/index.sh [hunk 16]
```diff
@@ -20,5 +20,11 @@ for meta in "${metas[@]}"; do
     # Ignore double quotes warning for yq expression
     # shellcheck disable=SC2016,SC2094
     cat <<< "$(yq --arg metadir "${META_DIR}" '.links |= . + {self: "/\($metadir)/devfile.yaml" }' "${meta}")"  > "${meta}"
+    if [ "$(yq '.links.v2' "${meta}")" != "null" ]; then
+      # shellcheck disable=SC2016,SC2094
```
Comment (benoitf) [inline build/scripts/index.sh:24]: I think it would be better to fix the issue rather than disabling the check

Comment (vinokurig) [reply build/scripts/index.sh:25]: The problem is that `yq` works with expressions in single quotes. We have the same approach above, added description.

**Comment by vinokurig** [comment]
@benoitf
> I tested but I have as links
>
>``` json
>"links": {
>"v2": "https://github.com/che-samples/java-guestbook/tree/devfilev2",
>"self": "/devfiles/java-mongo/devfile.yaml",
>"devWorkspaces": {
>"eclipse/che-theia/latest": "devfiles/java-mongo/devworkspace-che-theia-latest.yaml",
>"eclipse/che-theia/next": "devfiles/java-mongo/devworkspace-che-theia-next.yaml"
>}
>```
>
>so I suppose there is a missing / before devfiles ?

fixed

**Comment by vinokurig** [comment]
@nickboldt @svor @benoitf
> seems reasonable. Please remember to submit a PR for downstreaming this into https://github.com/redhat-developer/codeready-workspaces/tree/crw-2-rhel-8/dependencies/che-devfile-registry/devfiles for CRW 2.14

I think it should be done as a next iteration and in the scope of a separate issue.

**Review by benoitf — APPROVED** [review]
looks ok to me as first step

**Comment by benoitf** [comment]
probably it requires a change in release.sh to handle the tag replacement for latest images in DevWorkspaceTemplates but I don't know if you plan to do it now or later. But it needs to be adressed

**Comment by svor** [comment]
Release DevWorkspaceTemplates in devfile registry [#20824](https://github.com/eclipse/che/issues/20824)
Related issue for CRW: https://issues.redhat.com/browse/CRW-2500

**Comment by svor** [comment]
Is it ok that the content of resulted meta.yaml is in json format?

**Comment by benoitf** [comment]
@svor strange it should be a yaml file yes 

Probably there is a missing -y option

**Comment by svor** [comment]
@benoitf I mean in devfiles/go folder i see meta.yaml and its content is in JSON: 
```
bash-5.0# ls
devfile.yaml  devworkspace-che-theia-latest.yaml  devworkspace-che-theia-next.yaml  meta.yaml
bash-5.0# cat meta.yaml 
{
  "displayName": "Go",
  "description": "Stack with Go 1.14",
  "tags": [
    "Debian",
    "Go"
  ],
  "icon": "/images/go.svg",
  "links": {
    "v2": "https://github.com/che-samples/golang-echo-example/tree/devfile2",
    "self": "/devfiles/go/devfile.yaml",
    "devWorkspaces": {
      "eclipse/che-theia/latest": "/devfiles/go/devworkspace-che-theia-latest.yaml",
      "eclipse/che-theia/next": "/devfiles/go/devworkspace-che-theia-next.yaml"
    }
  }
}

```
But it's not related to this PR

**Review by svor — APPROVED** [review]

**Comment by nickboldt** [comment]
Downstream issue is https://issues.redhat.com/browse/CRW-2500

**Comment by benoitf** [comment]
@svor OK that's a bug that needs to be fixed

**Comment by svor** [comment]
> @svor OK that's a bug that needs to be fixed

issue for that https://github.com/eclipse/che/issues/20828

**Pull request merged**